### ISSUE TITLE
toSQL does not output "with clause" when defined inside a callback builder

### ISSUE THREAD

**Issue by ronyeh-lb** [knex/knex#5363]
# Environment

Knex version: 2.3.0
Database + version: psql 13.3
OS: ubuntu

Select applicable template from below.

# Bug

toSQL() **does not output** the "with" clause,when defined inside a builder, **altho it DOES submit it currently** with the with clause to the db

2. Error message

```
const Knex = require('knex');


const knexPg = Knex({
  client: 'pg',
});

(async function run() {

    let q = knexPg({ f: 'foo', b: 'bar' })
      .select('foo.*')
      .where('f.name', knexPg.raw('??', ['b.name']))
      .whereIn('something', knexPg('bar').select('id'))
          
    q.where(builder => {

        q.with('aa', knexPg.raw('select * from aa'));
})

  console.log(q.toSQL());
})();

```

current result

```
sql: "select \"foo\".* from \"foo\" as \"f\", \"bar\" as \"b\" …"name\" and \"something\" in (select \"id\" from \"bar\")"
```

**Comment by mercmobily** [comment]
Hi! Thanks for the clear repro. I think this is expected behaviour (though the error message could be clearer), not a `toSQL()` bug. 

When you pass a function to `where(() => { ... })`, Knex compiles that callback using a fresh inner QueryBuilder and the `whereWrapped` path. That inner builder is compiled with method `'where'`, producing only the `where ...` fragment for the outer query. Any `with()` you call on the inner builder won’t be emitted, because the `'where'` compiler method doesn’t include the `with` prefix, and only top‑level `select()` does.

It would help to add a short note in the CTE/joins/where sections of the documentation clarifying  where with() is effective. Specifically:

- Clarify that with() registers CTEs on the current (outer) builder and is emitted by the top‑level select; calling with() inside a where(() => {...}) or other nested callback compiles only the inner “where …” fragment and won’t emit WITH.
- Add a brief example showing the correct pattern (register with() on the outer builder) and, if needed, how to build a separate subquery with its own WITH and reference it from where().
- Maybe: link to whereWrapped behaviour to explain why inner with() calls don’t appear in toSQL().

Now, after some AI-aided searching:

- Inner compilation: `whereWrapped` uses `rawOrFn(..., 'where')` → `compileCallback` creates a new builder and compiles just the `where` portion (lib/query/querycompiler.js:1094–1120, lib/formatter/formatterUtils.js:1–18).
- Top‑level CTEs: the `with` prefix comes from the outer builder’s grouped `with` statements inside `select()` (lib/query/querycompiler.js:1206–1225).

What to do:

- Register CTEs on the outer builder, not inside the `where` callback:
  ```js
  const q = knex('foo as f')
    .with('aa', knex.raw('select * from bar'))
    .select('f.*')
    .where('f.name', knex.raw('??', ['b.name']));
  console.log(q.toSQL().sql); // includes WITH aa as (...)
  ```
- If you really need a CTE inside a subquery used in `where`, build that subquery separately (so it compiles as a full `select` with its own WITH) and reference it as a raw or a nested builder, rather than trying to register a CTE on the inner `where` callback builder.

In short: `with()` on the inner `where` callback’s builder is ignored in output by design; `with()` must be called on the outer builder to appear in `toSQL()` and in the final SQL. But, documentation should be improved.

References: lib/query/querycompiler.js:57–74, 1094–1120, 1206–1225; lib/formatter/formatterUtils.js:1–18
